README.md
# AtCoder Problems

[![Build Status](https://travis-ci.org/kenkoooo/AtCoderProblems.svg?branch=master)](https://travis-ci.org/kenkoooo/AtCoderProblems)

# atcoder-problems-backend

`atcoder-problems-backend` is a set of backend applications written in Rust.

## Crawler

### Build
```bash
cd atcoder-problems-backend/
cargo build --release --bin crawl
```
### Run
Before running, please make sure the URL of PostgreSQL is set to `SQL_URL`, and [SQL table schema](config/database-definition.sql) is loaded on your database.
```bash
export SQL_URL="postgresql://..." # URL of PostgreSQL database
cargo run --release --bin crawl [option]
```

#### Options
- `new_contests`
- `new`
- `all`
- `recent_submitted`
- `contest`

# atcoder-problems-frontend

`atcoder-problems-frontend` is a web application written in TypeScript.

## Install required packages
```bash
yarn
```

## Start the web application on your local
```bash
yarn start
```

## Build optimized files for distribution
```bash
yarn build
```

## Fix code format
```bash
yarn lint:fix
```

# Docker

```bash
#build Dockerfile
docker build ./

#run docker image
docker run -h spam -i -t {image_name} /bin/bash

#after you entered shell
#please start db server
service postgresql start

```
You can login db server with this user:name="kenkoooo",password="pass",dbname="test".
This docker image has everything you need for building and testing this repository(including prepared db server).
You can use git to get data and start testing.

# API

Please use API politely.
Especially, please ensure using cache and adding delay if you develop a bot.

APIは行儀良く利用してください。
特にbotを開発している場合は、ETagを用いたキャッシュを利用し、連続して呼び出す際はある程度 (例えば1秒) の遅延をはさむようにしてください。
1日に数万回以上のリクエストをしたい場合は事前に相談してください。

ブラウザ上からJavaScriptで直接利用している場合はあまり気にする必要はありません。
キャッシュはブラウザがよしなにやってくれますし、呼び出し回数もたいてい多くならないためです。
botの場合も1日に1000回以下程度のリクエストであればたいした負荷にはならないので心配する必要はありません。

## Information API

- Contests Information
  - https://kenkoooo.com/atcoder/resources/contests.json
- Problems Information
  - https://kenkoooo.com/atcoder/resources/problems.json
- Detailed Problems Information
  - https://kenkoooo.com/atcoder/resources/merged-problems.json
- Pairs of contests and problems
  - https://kenkoooo.com/atcoder/resources/contest-problem.json

(The API `info/*` are deprecated. Please use new API.)

## Statistics API

- Accepted Count
  - https://kenkoooo.com/atcoder/resources/ac.json
- Rated Point Sum
  - https://kenkoooo.com/atcoder/resources/sums.json
- Accepted Count for each language
  - https://kenkoooo.com/atcoder/resources/lang.json

## UserInfo API
- https://kenkoooo.com/atcoder/atcoder-api/v2/user_info?user=kenkoooo

## Submission API
- https://kenkoooo.com/atcoder/atcoder-api/results?user=wata

## Time-based Submission API
### Interface
```
https://kenkoooo.com/atcoder/atcoder-api/v3/from/{unix_time_second}
```
### Example
- https://kenkoooo.com/atcoder/atcoder-api/v3/from/1505342145

# Dataset

## Submissions

This will be updated once a week.

- https://s3-ap-northeast-1.amazonaws.com/kenkoooo/submissions.csv.gz

## Estimated difficulties of the problems
- https://kenkoooo.com/atcoder/resources/problem-models.json


CHANGELOG.md
# Change Log

## 2019-12-09
- [#343](https://github.com/kenkoooo/AtCoderProblems/pull/343) Upgraded `actix-web` to `2.0.0-alpha.4`
- [#342](https://github.com/kenkoooo/AtCoderProblems/pull/342) Fixed `grcov` command lines in the Travis config.
- [#341](https://github.com/kenkoooo/AtCoderProblems/pull/341) Fixed the Travis config to use the latest `grcov`.
- [#340](https://github.com/kenkoooo/AtCoderProblems/pull/340) Fixed es-lint warnings.
- [#339](https://github.com/kenkoooo/AtCoderProblems/pull/339) Configured the CORS header of the server response.
- [#338](https://github.com/kenkoooo/AtCoderProblems/pull/338) Added `fix_invalid_submissions`.
- [#337](https://github.com/kenkoooo/AtCoderProblems/pull/337) Added `variance` to the Problem Model API.

## 2019-12-07
- [#334](https://github.com/kenkoooo/AtCoderProblems/pull/334) Upgraded npm packages.
- [#332](https://github.com/kenkoooo/AtCoderProblems/pull/332) Added an `actix-web` based server application.

## 2019-12-05
- [#331](https://github.com/kenkoooo/AtCoderProblems/pull/331) Added `batch_update`

## 2019-11-21
- [#326](https://github.com/kenkoooo/AtCoderProblems/pull/326) Added a button to exclude experimental results.
- [#325](https://github.com/kenkoooo/AtCoderProblems/pull/325) Added a new SQL table to cache the result.

## 2019-11-16
- [#324](https://github.com/kenkoooo/AtCoderProblems/pull/324) Trimming of the string of the text form.

## 2019-11-01
- [#322](https://github.com/kenkoooo/AtCoderProblems/pull/322)  Added an option to hide experimental recommendations. 

## 2019-10-27
- [#320](https://github.com/kenkoooo/AtCoderProblems/pull/320) Refactoring of LanguageOwners.
- [#319](https://github.com/kenkoooo/AtCoderProblems/pull/319) Enabled to choose the number of shown language owners.
- [#318](https://github.com/kenkoooo/AtCoderProblems/pull/318) Difficulty estimation for old contests.

## 2019-10-24
- [#316](https://github.com/kenkoooo/AtCoderProblems/pull/316) Stop aggregating before-contest submissions.
- [#314](https://github.com/kenkoooo/AtCoderProblems/pull/314) Ported aggregation functions from Rust to SQL.

## 2019-10-23
- [#313](https://github.com/kenkoooo/AtCoderProblems/pull/313) Added `error::Error`.
- [#312](https://github.com/kenkoooo/AtCoderProblems/pull/312) Created SQL updater for AWS Lambda.
- [#310](https://github.com/kenkoooo/AtCoderProblems/pull/310) Removed unused AWS Lambda functions.

## 2019-10-20
- [#308](https://github.com/kenkoooo/AtCoderProblems/pull/308) Created a new Docker container for `backend`.
- [#307](https://github.com/kenkoooo/AtCoderProblems/pull/307) Compressed `submissions.csv`.

## 2019-10-19
- [#306](https://github.com/kenkoooo/AtCoderProblems/pull/306) Added ts-lint check into CI.
- [#305](https://github.com/kenkoooo/AtCoderProblems/pull/305) Removed unused python scripts.  

## 2019-10-18
- [#304](https://github.com/kenkoooo/AtCoderProblems/pull/304) Fixed the margins.
- [#303](https://github.com/kenkoooo/AtCoderProblems/pull/303) Installed ts-lint and Prettier.  

## 2019-10-15
- [#300](https://github.com/kenkoooo/AtCoderProblems/pull/300) Removed corrupted difficulty.
- [#299](https://github.com/kenkoooo/AtCoderProblems/pull/299) Updated `Docker.crawler` to reduce the image size.

## 2019-10-14
- [#297](https://github.com/kenkoooo/AtCoderProblems/pull/297) Updated `Docker.crawler` to fix the compile errors in DockerHub.
- [#296](https://github.com/kenkoooo/AtCoderProblems/pull/296) Updated `rusoto` to fix the compile errors.
- [#294](https://github.com/kenkoooo/AtCoderProblems/pull/294) Updated the difficulty estimation algorithm.
- [#293](https://github.com/kenkoooo/AtCoderProblems/pull/293) Updated the sql backup script to compress the CSV file.
- [#291](https://github.com/kenkoooo/AtCoderProblems/pull/291) Updated crawler to handle DB connection errors.


main.go
package main

import (
	"database/sql"
	"flag"
	"fmt"
	sq "github.com/Masterminds/squirrel"
	"github.com/PuerkitoBio/goquery"
	"github.com/Sirupsen/logrus"
	_ "github.com/go-sql-driver/mysql"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

type Submit struct {
	Id           int
	ProblemId    string
	User         string
	Language     string
	SourceLength int
	Status       string
	ExecTime     int
	CreatedAt    string
}

func (s *Submit) IdStr() string {
	return strconv.Itoa(s.Id)
}

func GetContestUrls() []string {
	x := []string{}
	doc, _ := goquery.NewDocument("http://atcoder.jp/")
	rep := regexp.MustCompile(`^https*://([a-z0-9\-]*)\.contest\.atcoder.*$`)
	doc.Find("a").Each(func(_ int, s *goquery.Selection) {
		url, _ := s.Attr("href")
		if rep.Match([]byte(url)) {
			url = rep.ReplaceAllString(url, "$1")
			x = append(x, url)
		}
	})
	return x
}

func GetProblemSet(contest string) []string {
	set := make(map[string]bool)
	url := "http://" + contest + ".contest.atcoder.jp/assignments"
	doc, _ := goquery.NewDocument(url)
	rep := regexp.MustCompile(`^/tasks/([0-9_a-z]*)$`)
	doc.Find("a").Each(func(_ int, s *goquery.Selection) {
		url, _ := s.Attr("href")
		if rep.Match([]byte(url)) {
			url = rep.ReplaceAllString(url, "$1")
			set[url] = true
		}
	})

	x := []string{}
	for key, _ := range set {
		x = append(x, key)
	}
	return x
}

func GetSubmissions(contest string, i int) ([]Submit, int) {
	x := []Submit{}
	max := 1

	url := "http://" + contest + ".contest.atcoder.jp/submissions/all/" + strconv.Itoa(i)
	doc, _ := goquery.NewDocument(url)

	allrep := regexp.MustCompile(`^/submissions/all/([0-9]*)$`)
	doc.Find("a").Each(func(_ int, s *goquery.Selection) {
		url, _ := s.Attr("href")
		if allrep.Match([]byte(url)) {
			url = allrep.ReplaceAllString(url, "$1")
			page, _ := strconv.Atoi(url)
			if page > max {
				max = page
			}
		}
	})

	rep := regexp.MustCompile(`^/submissions/([0-9]*)$`)
	prep := regexp.MustCompile(`^/tasks/([0-9_a-z]*)$`)
	urep := regexp.MustCompile(`^/users/([0-9_a-zA-Z]*)$`)

	doc.Find("tbody").Each(func(_ int, s *goquery.Selection) {
		s.Find("tr").Each(func(_ int, s *goquery.Selection) {
			var key int
			var problem_id string
			var user_name string
			s.Find("a").Each(func(_ int, t *goquery.Selection) {
				url, _ := t.Attr("href")
				if rep.Match([]byte(url)) {
					url = rep.ReplaceAllString(url, "$1")
					key, _ = strconv.Atoi(url)
				} else if prep.Match([]byte(url)) {
					problem_id = prep.ReplaceAllString(url, "$1")
				} else if urep.Match([]byte(url)) {
					user_name = urep.ReplaceAllString(url, "$1")
				}
			})

			data := []string{}
			s.Find("td").Each(func(_ int, s *goquery.Selection) {
				data = append(data, s.Text())
			})
			length, _ := strconv.Atoi(strings.Replace(data[5], " Byte", "", -1))
			t := Submit{
				Id:           key,
				ProblemId:    problem_id,
				User:         user_name,
				Language:     data[3],
				SourceLength: length,
				Status:       data[6],
				CreatedAt:    data[0],
				ExecTime:     0,
			}
			if len(data) == 10 {
				exec_time, _ := strconv.Atoi(strings.Replace(data[7], " ms", "", -1))
				t.ExecTime = exec_time
			}
			x = append(x, t)
		})
	})

	return x, max
}

func GetMySQL(user string, pass string) (db *sql.DB) {
	server := user + ":" + pass + "@/atcoder"
	db, err := sql.Open("mysql", server)
	if err != nil {
		panic(err.Error())
	}
	return db
}

func NewRecord(table, column, key string, db *sql.DB) bool {
	query, args, _ := sq.Select(column).From(table).Where(sq.Eq{column: key}).Limit(1).ToSql()
	row, _ := db.Query(query, args...)
	defer row.Close()
	for row.Next() {
		return false
	}
	return true
}

func UpdateProblemSet(db *sql.DB, logger *logrus.Logger) {
	urls := GetContestUrls()
	for _, contest := range urls {
		if !NewRecord("contests", "id", contest, db) {
			continue
		}

		problems := GetProblemSet(contest)
		if len(problems) == 0 {
			continue
		}
		logger.WithFields(logrus.Fields{"contest": contest}).Info("crawling problems")
		query, args, _ := sq.Insert("contests").Columns("id").Values(contest).ToSql()
		db.Exec(query, args...)
		q := sq.Insert("problems").Columns("id", "contest")
		for _, problem := range problems {
			if NewRecord("problems", "id", problem, db) {
				q = q.Values(problem, contest)
			}
		}
		query, args, _ = q.ToSql()
		db.Exec(query, args...)
	}
}

func UpdateSubmissions(db *sql.DB, logger *logrus.Logger) {
	contest := ""
	{
		query, args, _ := sq.Select("id").From("contests").OrderBy("last_crawled").Limit(1).ToSql()
		row, _ := db.Query(query, args...)
		for row.Next() {
			row.Scan(&contest)
			fmt.Println(contest)
		}
	}

	M := 1
	for i := 1; i <= M; i++ {
		submissions, max := GetSubmissions(contest, i)
		logger.WithFields(logrus.Fields{"contest": contest, "page": strconv.Itoa(i)}).Info("crawling page")
		if max > M {
			M = max
		}
		q := sq.Insert("submissions")
		q = q.Columns(
			"id", "problem_id", "contest_id",
			"user_name", "status", "source_length", "language",
			"exec_time", "created_time")

		crawled_flag := false
		for _, s := range submissions {
			if NewRecord("submissions", "id", s.IdStr(), db) {
				q = q.Values(
					s.Id, s.ProblemId, contest,
					s.User, s.Status, s.SourceLength, s.Language,
					s.ExecTime, s.CreatedAt)
			} else {
				crawled_flag = true
			}
		}
		query, args, _ := q.ToSql()
		db.Exec(query, args...)
		fmt.Println(i)
		if crawled_flag {
			break
		}
	}

	query, args, _ := sq.Update("contests").Set("last_crawled", time.Now().Format("2006-01-02 15:04:05")).Where(sq.Eq{"id": contest}).ToSql()
	db.Exec(query, args...)
}

const cycle = 24 * 3600
const onesec_nano = 1000000000

func main() {
	u := flag.String("u", "user", "user name to connect to MySQL server")
	p := flag.String("p", "password", "password to connect to MySQL server")
	flag.Parse()

	db := GetMySQL(*u, *p)
	defer db.Close()

	logger := logrus.New()
	logger.Formatter = new(logrus.JSONFormatter)
	if err := os.Mkdir("log", 0777); err != nil {
		fmt.Println(err)
	}

	for i := 0; ; i++ {
		i %= cycle

		f, _ := os.OpenFile("log/"+time.Now().Format("2006-01-02")+".log", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
		logger.Out = f
		if i == 0 {
			UpdateProblemSet(db, logger)
		} else {
			UpdateSubmissions(db, logger)
		}
		f.Close()

		time.Sleep(onesec_nano)
	}
}
